File: web-client/src/api/reportingApi.ts
import apiClient from './client';
import type { MyTaskReport, TaskHoursRow, ProjectHoursRow, DetailedHoursRow, ProjectTaskCountRow } from './types';

/** Fetch open tasks assigned to the current user (RELEASED/REJECTED excluded). */
export function getMyTasks(): Promise<MyTaskReport[]> {
  return apiClient.get<MyTaskReport[]>('/api/v1/reports/my-tasks').then((r) => r.data);
}

/** Fetch open tasks assigned to the current user updated within the last {@code days} days. */
export function getMyTasksFiltered(days: number): Promise<MyTaskReport[]> {
  return apiClient.get<MyTaskReport[]>(`/api/v1/reports/my-tasks?days=${days}`).then((r) => r.data);
}

/** Fetch finished tasks (RELEASED or REJECTED) assigned to the current user. */
export function getMyFinishedTasks(): Promise<MyTaskReport[]> {
  return apiClient.get<MyTaskReport[]>('/api/v1/reports/my-tasks?finished=true').then((r) => r.data);
}

/** Fetch planned vs booked hours per task, optionally filtered by project. */
export function getHoursByTask(projectId?: string): Promise<TaskHoursRow[]> {
  const params = projectId ? `?projectId=${projectId}` : '';
  return apiClient.get<TaskHoursRow[]>(`/api/v1/reports/hours/by-task${params}`).then((r) => r.data);
}

/** Fetch planned vs booked hours totalled per project. */
export function getHoursByProject(): Promise<ProjectHoursRow[]> {
  return apiClient.get<ProjectHoursRow[]>('/api/v1/reports/hours/by-project').then((r) => r.data);
}

/** Fetch planned vs booked hours for a task broken down by user and work type. */
export function getHoursDetailed(taskId: string): Promise<DetailedHoursRow[]> {
  return apiClient.get<DetailedHoursRow[]>(`/api/v1/reports/hours/detailed?taskId=${taskId}`).then((r) => r.data);
}

/** Fetch open task counts per project split by total and the current user's assigned count. */
export function getOpenTasksByProject(): Promise<ProjectTaskCountRow[]> {
  return apiClient.get<ProjectTaskCountRow[]>('/api/v1/reports/tasks/open-by-project').then((r) => r.data);
}

File: android/data/src/main/kotlin/com/demo/taskmanager/data/api/ReportingApi.kt
package com.demo.taskmanager.data.api

import com.demo.taskmanager.data.dto.HoursByProjectDto
import com.demo.taskmanager.data.dto.HoursByTaskDto
import com.demo.taskmanager.data.dto.HoursDetailedDto
import com.demo.taskmanager.data.dto.MyTaskReportDto
import com.demo.taskmanager.data.dto.ProjectOpenTaskCountDto
import retrofit2.http.GET
import retrofit2.http.Query

/** Retrofit interface for reporting-service: task and hours reports. */
interface ReportingApi {

    /** Returns tasks assigned to or participated in by the current user, optionally filtered by look-ahead days. */
    @GET("api/v1/reports/my-tasks")
    suspend fun getMyTasks(@Query("days") days: Int? = null): List<MyTaskReportDto>

    /** Planned vs booked hours per task within a project. */
    @GET("api/v1/reports/hours/by-task")
    suspend fun getHoursByTask(@Query("projectId") projectId: String): List<HoursByTaskDto>

    /** Planned vs booked hours aggregated per project. */
    @GET("api/v1/reports/hours/by-project")
    suspend fun getHoursByProject(): List<HoursByProjectDto>

    /** Planned vs booked hours for a task broken down by user and work type. */
    @GET("api/v1/reports/hours/detailed")
    suspend fun getHoursDetailed(@Query("taskId") taskId: String): List<HoursDetailedDto>

    /** Open task counts per project — personal and project-wide totals. */
    @GET("api/v1/reports/tasks/open-by-project")
    suspend fun getOpenTasksByProject(): List<ProjectOpenTaskCountDto>
}
